// @celo-tools/use-contractkit/lib/use-contractkit.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useContractKitInternal = exports.useContractKit = void 0;
var contract_kit_provider_1 = require("./contract-kit-provider");
var useContractKit = function () {
    var _a = contract_kit_provider_1.useContractKitContext(), _b = _a[0], dapp = _b.dapp, connector = _b.connector, connectorInitError = _b.connectorInitError, address = _b.address, network = _b.network, _dispatch = _a[1], _c = _a[2], destroy = _c.destroy, updateNetwork = _c.updateNetwork, connect = _c.connect, getConnectedKit = _c.getConnectedKit, performActions = _c.performActions;
    return {
        address: address,
        dapp: dapp,
        network: network,
        updateNetwork: updateNetwork,
        kit: connector.kit,
        walletType: connector.type,
        account: connector.account,
        initialised: connector.initialised,
        performActions: performActions,
        getConnectedKit: getConnectedKit,
        connect: connect,
        destroy: destroy,
        initError: connectorInitError,
    };
};
exports.useContractKit = useContractKit;
/**
 * useContractKit with internal methods exposed. Package use only.
 */
var useContractKitInternal = function () {
    var _a = contract_kit_provider_1.useContractKitContext(), _b = _a[0], pendingActionCount = _b.pendingActionCount, connectionCallback = _b.connectionCallback, _dispatch = _a[1], initConnector = _a[2].initConnector;
    return __assign(__assign({}, exports.useContractKit()), { connectionCallback: connectionCallback, initConnector: initConnector, pendingActionCount: pendingActionCount });
};
exports.useContractKitInternal = useContractKitInternal;
